src/bin/offline/dl.rs

```diff
@@ -91,7 +96,19 @@ pub fn download(
 
         debug!("Compressing...");
 
-        offline::fs::save_compressed(&save_path, &data)?;
+        // When the pool is busy this compresses on the main thread, to avoid using a huge amount
+        // of RAM to store all the queued logs.
+        if compression_pool.active_count() >= compression_pool.max_count() {
+            debug!("compression pool is busy, compressing on the main thread...");
+            offline::fs::save_compressed(&save_path, &data)?;
+        } else {
+            debug!("compressing on the pool...");
+            compression_pool.execute(move || {
+                crate::util::log_and_exit_error(move || {
+                    offline::fs::save_compressed(&save_path, &data)
+                });
+            });
+        }
```

Comment:
This seems fine, though a bit low-level. I guess rayon and a for_each method won't work? Have you tried that?

Reply:
Yep, I tried rayon, but it didn't like trait objects much.

Reply:
Huh. I would've expected `+ Send + Sync` on the trait object to be sufficient.

Reply:
(anyway don't treat this as blocking)